src/goto-symex/show_program.cpp

```diff
@@ -78,3 +84,243 @@ void show_program(const namespacet &ns, const symex_target_equationt &equation)
       show_step(ns, step, "SHARED_WRITE", count);
   }
 }
+
+bool duplicated_previous_step(const SSA_stept &ssa_step)
+{
+  return !(
+    ssa_step.is_assignment() || ssa_step.is_assert() || ssa_step.is_assume() ||
+    ssa_step.is_constraint() || ssa_step.is_shared_read() ||
+    ssa_step.is_shared_write());
+}
+
+void show_ssa_step_plain(
+  messaget::mstreamt &out,
+  const namespacet &ns,
+  const SSA_stept &ssa_step,
+  const exprt &ssa_expr)
+{
+  const irep_idt &function_id = ssa_step.source.function_id;
+  const std::string ssa_expr_as_string = from_expr(ns, function_id, ssa_expr);
+
+  out << messaget::faint << "// " << ssa_step.source.pc->location_number << " ";
+  out << ssa_step.source.pc->source_location.as_string() << "\n"
+      << messaget::reset;
+  out << ssa_expr_as_string << "\n";
+}
+
+void show_ssa_step_json(
+  std::ostream &out,
+  const namespacet &ns,
+  const SSA_stept &ssa_step,
+  const exprt &ssa_expr,
+  json_arrayt &byte_op_list)
+{
+  const std::string key_srcloc = "sourceLocation";
+  const std::string key_ssa_expr = "ssaExpr";
+  const std::string key_ssa_expr_as_string = "ssaExprString";
+
+  json_objectt &json_ssa_step = byte_op_list.push_back(jsont()).make_object();
```

Comment:
:pick: The way you are creating this is a little mutation heavy. I'd prefer if the `json_objectt` didn't need to be mutated after it was constructed.
:bulb: The `json_objectt` class supports list initialisation. For example the `show_ssa_step_json` could be written like this -
```
void show_ssa_step_json(
  const namespacet &ns,
  const SSA_stept &ssa_step,
  const exprt &ssa_expr,
  json_arrayt &byte_op_list)
{
  const irep_idt &function_id = ssa_step.source.function_id;
  const std::string ssa_expr_as_string = from_expr(ns, function_id, ssa_expr);
  json_objectt json_ssa_step{
    {"sourceLocation", json(ssa_step.source.pc->source_location)},
    {"ssaExprString", json_stringt(ssa_expr_as_string)},
    {"ssaExpr", json_irept(false).convert_from_irep(ssa_expr)}};
  byte_op_list.push_back(std::move(json_ssa_step));
}
```
:information_source: Note that classes derived from `jsont` can be handled/copied by value. This does not risk object slicing with this particular hierarchy. This is because all the fields are defined in `jsont` and none of the derived classes add extra fields.

Reply:
Comment addressed. Pushed changes.